Keepassx fails to compile on FreeBSD due to libusb implementation differences

madpilot78 opened the pull request:
On FreeBSD the libusb_hotplug_register_callback() function uses a pointer to a struct as an handle.

I'm changing the methods to use this as expected by the system library.

This now compiles fine but fails at runtime.

Not sure why this patch is not working and I must admit I do not understand what is being done using `qintptr` in https://github.com/keepassxreboot/keepassxc/blob/f0932914424543f29e81f6c8933bbe2e8553b68c/src/gui/osutils/DeviceListener.h#L41.

This has been noticed here: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=277652

The failures with my patch are described here: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=277652#c10

I do not have a yubikey so I cannot do testing myself.

Thanks in advance for any help!


## Testing strategy

As said, with this changes the source compiles on FreeBSD, and the resulting binary works, except when exercising the modified code.

## Type of change
- ✅ Bug fix (non-breaking change that fixes an issue)

phoerious reviewed (commented):
Instead of adding all the ifdefs, you should cast libusb_hotplug_callback_handle to qintptr instead. This should already be an opaque pointer type that needs no additional logic.

phoerious commented on src/gui/osutils/nixutils/DeviceListenerLibUsb.cpp:
```diff
@@ -102,7 +111,7 @@ int DeviceListenerLibUsb::registerHotplugCallback(bool arrived, bool left, int v
     return handle;
 }
 
-void DeviceListenerLibUsb::deregisterHotplugCallback(int handle)
```
This is actually a bug and should be qintptr aka Handle.

phoerious commented on src/gui/osutils/nixutils/DeviceListenerLibUsb.h:
```diff
@@ -36,16 +43,17 @@ class DeviceListenerLibUsb : public QObject
 
 private:
     void* m_ctx;
-    QSet<int> m_callbackHandles;
```
Same here.

madpilot78 commented:
@phoerious Thanks for the feedback!

So, if I understand correctly, there is an existing bug in the code that uses `int`s where it should be using qintptr. This makes some sense to me, now I can understand why qintptr was used to start with.

I'll do some testing hoping I can make sense of all this.

Thanks again for your suggestions.

phoerious commented:
On a normal 64 bit system it shouldn't make a difference, but with smaller word lengths it may.

madpilot78 commented:
@phoerious Problem is, the code as is fails to compile on FreeBSD. FreeBSD is using recent clang by default, which tends to be less permissive about this kind of casts.

Putting the correct types in the correct place should be enough to make it compile and work then.

madpilot78 committed:
WIP: Change type of Handle on FreeBSD.

On FreeBSD the libusb_hotplug_register_callback() function uses a
pointer to a struct as an handle.

I'm changing the methods to use this as expected by the system library.

This now compiles fine but fails at runtime.

madpilot78 commented:
Hi again,

I experimented a little but made little progress.

I uploaded what I have now, following suggestion to use `qintptr`.

This fails to compile. If I use classic C style casts in place of static_cast it does compile but fails at runtime.

I'm asking for help on how to fix this. My limited C++ knowledge has already reached its limit.

Thanks in advance.

droidmonkey commented on src/gui/osutils/nixutils/DeviceListenerLibUsb.cpp:
```diff
@@ -49,7 +49,7 @@ namespace
 } // namespace
 
-int DeviceListenerLibUsb::registerHotplugCallback(bool arrived, bool left, int vendorId, int productId, const QUuid*)
+qintptr DeviceListenerLibUsb::registerHotplugCallback(bool arrived, bool left, int vendorId, int productId, const QUuid*)
```
This should be `Handle`

droidmonkey commented on src/gui/osutils/nixutils/DeviceListenerLibUsb.cpp:
```diff
@@ -84,7 +84,7 @@ int DeviceListenerLibUsb::registerHotplugCallback(bool arrived, bool left, int v
         },
         that,
-        &handle);
+        &static_cast<libusb_hotplug_callback_handle>(handle));
```
Both your build and runtime error is here, you shouldn't reference the output of the static cast directly. Make the handle variable an int, then static cast the return call.

```
error: cannot take the address of an rvalue of type 'libusb_hotplug_callback_handle' (aka 'int')
        &static_cast<libusb_hotplug_callback_handle>(handle));
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

madpilot78 replied:
Why int?

Please keep in mind that I'm trying to fix this in FreeBSD. FreeBSD has its own implementation of libusb. AFAIK other BSDs do too, but I have not checked.

FreeBSD's `libusb_hotplug_callback_handle` is defined as [1]:

```c
typedef struct libusb_hotplug_callback_handle_struct *libusb_hotplug_callback_handle;
```

so the error I'm getting is:

```
/wrkdirs/usr/ports/security/keepassxc/work/keepassxc-2.7.8/src/gui/osutils/nixutils/DeviceListenerLibUsb.cpp:87:10: error: cannot cast from type 'Handle' (aka 'long long') to pointer type 'libusb_hotplug_callback_handle' (aka 'libusb_hotplug_callback_handle_struct *')
   87 |         &static_cast<libusb_hotplug_callback_handle>(handle));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/wrkdirs/usr/ports/security/keepassxc/work/keepassxc-2.7.8/src/gui/osutils/nixutils/DeviceListenerLibUsb.cpp:110:77: error: cannot cast from type 'Handle' (aka 'long long') to pointer type 'libusb_hotplug_callback_handle' (aka 'libusb_hotplug_callback_handle_struct *')
  110 |     libusb_hotplug_deregister_callback(static_cast<libusb_context*>(m_ctx), static_cast<libusb_hotplug_callback_handle>(handle));
      |                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
```

I'm not sure what I should cast to at this point, on FreeBSD.

If later Linux (or any other OS) requires a different code I will use `#ifdef`s to differentiate.


[1] https://github.com/freebsd/freebsd-src/blob/48edad2edf6eb7a539e40dad8e1f87e3fa4973fd/lib/libusb/libusb.h#L311

droidmonkey replied:
Either way your static cast is incorrect in this context. Slow clap for FreeBSD for a completely divergent API with the same exact function name.

madpilot78 replied:
Well I don't know how and why a different API evolved in FreeBSD, but I'm trying to understand why a 64bit number on a 64bit arch cannot be cast to a pointer. The memory representation of them is the same AFAIK.

What I'm trying to do is adapt the code in keepassxc for these function to pass around pointers in place of integers. Apart fro the different type (with a different size in most cases), the API is not that much different, so the basic concept should be possible to put in practice.

phoerious replied:
On Ubuntu (libusb 1.0.26), `libusb_hotplug_callback_handle` is typedef'ed to `int`. FreeBSD has an entirely different implementation of the whole library.

I think the main problem is that `libusb_hotplug_callback_handle` itself is not a pointer type on FreeBSD. Instead, they chose to typedef only `*libusb_hotplug_callback_handle`, so you have to use it with `*`.

madpilot78 replied:
@phoerious

Looks like FreeBSD has an old implementation of libusb included in the base system. This will be addressed at some point, but will require time.

I can try a different patch, but since I do not own a ubikey I can properly test them myself, and getting feedback requires time.

madpilot78 committed:
Correct return type.

droidmonkey commented:
@phoerious what do you want to do here?

madpilot78 commented:
For completeness, these bugs have been created regarding FreeBSD:

https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=277799

https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=279556

But they apply to base system and resolution cannot be fast, most probably nothing can get released before the next major release at the earliest.

phoerious committed:
Fix Linux/FreeBSD incompatibility

phoerious commented:
I've pushed a commit that should make it work on both Linux and FreeBSD, though I cannot test the latter. If it works, we should rebase and squash the commits.

Do you know whether any other *BSD will be affected as well? I checked OpenBSD, but they don't seem to have a custom libusb in their tree.

droidmonkey commented:
Looks reasonable to me

madpilot78 commented:
@phoerious Thanks a lot for your patch!

I've compiled and am running keepassxc with it locally, it works fine, but I am unable to test the USB functionality.

I've posted a patch against the ports tree including your patch so others can test on FreeBSD, I hope to have feedback in a few days at most.

phoerious commented:
You should be able to test it to some degree even without a YubiKey. Open two tabs and switch between them. You should see a green progress bar appear for a short time below the password field. If that's the case and the app doesn't crash, it most likely works.

madpilot78 commented:
@phoerious

I performed a quick test you suggest by creating a new DB. Unluckily changing tabs causes a lockup of the program, I am unable to do anything else. I need to kill it with `xkill`.

phoerious commented:
That is strange, though probably not related to this change. It may, however, be a problem with that libusb implementation.
